module/zfs/zfs_sa.c

```diff
@@ -264,6 +278,8 @@ zfs_sa_set_xattr(znode_t *zp)
 		VERIFY0(sa_bulk_update(zp->z_sa_hdl, bulk, count, tx));
 
 		dmu_tx_commit(tx);
+		if (logsaxattr && zfsvfs->z_os->os_sync == ZFS_SYNC_ALWAYS)
```

Comment:
Why do we only `zil_commit` for `ZFS_SYNC_ALWAYS`?

I guess because sync semantics of xattrs are not defined? I feel like this should be a separate filesystem property, like, `xattr_sync=on|off`.

Such a property would also make the `zfs_zil_saxattr_enabled` tunable obsolete (if `xattr_sync=off`, act as if the tunable was off).

Reply:
Yes. Specific sync semantics are not defined for xattr's.
In case of xattr=dir,  zil_commit() is done either when ZFS_SYNC_ALWAYS true on objset OR file opened with (O_SYNC | O_DSYNC).  As here, we only have control to check on ZFS_SYNC_ALWAYS for now, so doing that only.
In future, it can possibly be extended to do have sync control at file level similar to xattr=dir.